server/src/main/java/org/elasticsearch/action/admin/cluster/stats/CCSTelemetrySnapshot.java

```diff
@@ -0,0 +1,376 @@
+        stats.featureCounts.forEach((k, v) -> featureCounts.merge(k, v, Long::sum));
+        stats.clientCounts.forEach((k, v) -> clientCounts.merge(k, v, Long::sum));
+        took.add(stats.took);
+        tookMrtTrue.add(stats.tookMrtTrue);
+        tookMrtFalse.add(stats.tookMrtFalse);
+        remotesPerSearchMax = Math.max(remotesPerSearchMax, stats.remotesPerSearchMax);
+        if (totalCount > 0 && oldCount > 0) {
+            // Weighted average
+            remotesPerSearchAvg = (remotesPerSearchAvg * oldCount + stats.remotesPerSearchAvg * stats.totalCount) / totalCount;
+        } else {
+            // If we didn't have any old value, we just take the new one
+            remotesPerSearchAvg = stats.remotesPerSearchAvg;
+        }
+        // we copy the object here since we'll be modifying it later on subsequent adds
+        // TODO: this may be sub-optimal, as we'll be copying histograms when adding first snapshot to an empty container,
+        // which we could have avoided probably.
+        stats.byRemoteCluster.forEach((r, v) -> byRemoteCluster.merge(r, new PerClusterCCSTelemetry(v), PerClusterCCSTelemetry::add));
+    }
+
+    /**
+     * Publishes the latency statistics to the provided {@link XContentBuilder}.
+     * Example:
+     * "took": {
+     *      "max": 345032,
+     *      "avg": 1620,
+     *      "p90": 2570
+     * }
+     */
+    public static void publishLatency(XContentBuilder builder, String name, LongMetricValue took) throws IOException {
+        builder.startObject(name);
+        {
+            builder.field("max", took.max());
+            builder.field("avg", took.avg());
+            builder.field("p90", took.p90());
+        }
+        builder.endObject();
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject("ccs_telemetry");
+        {
+            builder.field("total", totalCount);
+            builder.field("success", successCount);
+            builder.field("skipped", skippedRemotes);
+            publishLatency(builder, "took", took);
+            publishLatency(builder, "took_mrt_true", tookMrtTrue);
+            publishLatency(builder, "took_mrt_false", tookMrtFalse);
+            builder.field("remotes_per_search_max", remotesPerSearchMax);
+            builder.field("remotes_per_search_avg", remotesPerSearchAvg);
+            builder.field("failure_reasons", failureReasons);
+            builder.field("features", featureCounts);
+            builder.field("clients", clientCounts);
+            builder.startObject("remote_clusters");
+            {
+                builder.field("count", byRemoteCluster.size());
+                byRemoteCluster.forEach((name, clusterData) -> {
+                    try {
+                        builder.field(name, clusterData);
+                    } catch (IOException e) {
```

Comment:
Asking to learn - why do we need this catch block? And if it gets an IOException, what is the consequence of throwing a RuntimeException? We don't want to crash a server over this issue.

Reply:
Because it's a lambda, and lambdas in Java can't have checked exceptions. So if it's not caught, Java complains about lambda missing checked exception (which `IOException` is) so it has to be converted to runtime. Lambdas and checked exceptions don't play nice unfortunately. I could unroll it to for loop if it feels nicer (I like `forEach` better but it's just personal preference).

Reply:
The probability it will actually throw in this particular line is minimal, there's not much there to cause it, but I don't have much choice since `field` is declared with checked exception.